### ISSUE TITLE
How did you solve the issue with bcrypt, if I may ask?

### ISSUE THREAD

**Issue by tobkle** [technomancy-dev/00#21]
When I do fly launch it will compile everything.
Haven't used Erlang before. However, I have installed now:
* OSX Sonoma M1 Max
* zsh or bash: same results
* Erlang/OTP 26
* mix Erlang/OTP 26
* gcc Apple clang version 15.0.0 (clang-1500.3.9.4) Target: arm64-apple-darwin23.5.0
* make GNU Make 3.81

But I have no idea how to get this bcrypt_elixir thing compiled. The mentioned commands below didn't work:
* "mix deps.compile bcrypt_elixir --force"
* "mix deps.update bcrypt_elixir" 
* "mix deps.clean bcrypt

```zsh
All dependencies are up to date
==> bcrypt_elixir
mkdir -p "/Users/xxx/code/mail-host/00/_build/dev/lib/bcrypt_elixir/priv"
cc -g -O3 -Wall -Wno-format-truncation -I"/opt/homebrew/Cellar/erlang/26.2.5/lib/erlang/erts-14.2.5/include" -Ic_src -fPIC -shared -L/opt/homebrew/opt/llvm/lib
clang: error: no input files
make: *** [_lib_name] Error 1
could not compile dependency :bcrypt_elixir, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile bcrypt_elixir --force", update it with "mix deps.update bcrypt_elixir" or clean it with "mix deps.clean bcrypt_elixir"
==> phoenix_00
** (Mix) Could not compile with "make" (exit status: 2).
You need to have gcc and make installed. Try running the
commands "gcc --version" and / or "make --version". If these programs
are not installed, you will be prompted to install them.
```

**Comment by liltechnomancer** [comment]
Hey! The project actually uses elixir which is built on top of erlang so you would likely need that to fix the error. https://elixir-lang.org/install.html

However, fly should have wanted to just deploy the docker container and you shouldn't require you install anything. Are you trying to run it locally? The easiest way to do that is also with docker.

**Comment by tobkle** [comment]
Thanks. Yes I had also Elixir setup. If I do ```fly launch``` it tries to compile the Erlang/Phoenix sources.
Had started the Docker Container also directly with the .env entries, but there was not any port answering.
So I tried to compile it myself. So far I didn't get it working. Keep trying. ;-) Regards Toby

**Comment by liltechnomancer** [comment]
The issue with fly launch may be that the fly.toml image already exists. Might need to delete that before running. @tobkle 

For running locally you should tell it to listen on `4000` like `docker run -it --env-file .env -p 4000:4000 "liltechnomancer/double-zero"`

**Comment by liltechnomancer** [comment]
Hey @tobkle are you still experiencing issues?

**Comment by tobkle** [comment]
Yes. Cannot get it working.

If I do fly launch, I'm getting these errors:
```sh
All dependencies are up to date
==> bcrypt_elixir
mkdir -p "/Users/toby/code/mail-host/01/_build/dev/lib/bcrypt_elixir/priv"
cc -g -O3 -Wall -Wno-format-truncation -I"/opt/homebrew/Cellar/erlang/26.2.5/lib/erlang/erts-14.2.5/include" -Ic_src -fPIC -shared -L/opt/homebrew/opt/llvm/lib
clang: error: no input files
make: *** [_lib_name] Error 1
could not compile dependency :bcrypt_elixir, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile bcrypt_elixir --force", update it with "mix deps.update bcrypt_elixir" or clean it with "mix deps.clean bcrypt_elixir"
==> phoenix_00
** (Mix) Could not compile with "make" (exit status: 2).
You need to have gcc and make installed. Try running the
commands "gcc --version" and / or "make --version". If these programs
are not installed, you will be prompted to install them.
```

If I run the original docker image locally, I do get the following errors, as soon as I come to localhost:4000:
```sh
2024-08-04 19:25:24 17:25:24.341 [error] Could not check origin for Phoenix.Socket transport.
2024-08-04 19:25:24 
2024-08-04 19:25:24 Origin of the request: http://localhost:4000
2024-08-04 19:25:24 
2024-08-04 19:25:24 This happens when you are attempting a socket connection to
2024-08-04 19:25:24 a different host than the one configured in your config/
2024-08-04 19:25:24 files. For example, in development the host is configured
2024-08-04 19:25:24 to "localhost" but you may be trying to access it from
2024-08-04 19:25:24 "127.0.0.1". To fix this issue, you may either:
2024-08-04 19:25:24 
2024-08-04 19:25:24   1. update [url: [host: ...]] to your actual host in the
2024-08-04 19:25:24      config file for your current environment (recommended)
2024-08-04 19:25:24 
2024-08-04 19:25:24   2. pass the :check_origin option when configuring your
2024-08-04 19:25:24      endpoint or when configuring the transport in your
2024-08-04 19:25:24      UserSocket module, explicitly outlining which origins
2024-08-04 19:25:24      are allowed:
2024-08-04 19:25:24 
2024-08-04 19:25:24         check_origin: ["https://example.com",
2024-08-04 19:25:24                        "//another.com:888", "//other.com"]
2024-08-04 19:25:24 
2024-08-04 19:25:24 
2024-08-04 19:25:24 17:25:24.345 [notice] SIGTERM received - shutting down
2024-08-04 19:25:24 
2024-08-04 19:25:24 17:25:24.362 [info] CONNECTED TO Phoenix.LiveView.Socket in 152µs
2024-08-04 19:25:24   Transport: :longpoll
2024-08-04 19:25:24   Serializer: Phoenix.Socket.V2.JSONSerializer
2024-08-04 19:25:24   Parameters: %{"_csrf_token" => "AjwhGjE2b2xBf3g7Cjs4J0onBHEGcjkslFkbYg9Xp2JteANIsNMC7_XH", "_live_referer" => "undefined", "_mounts" => "0", "_track_static" => %{"0" => "http://localhost:4000/assets/app-f31482054921925fd908594c96a609e0.css?vsn=d", "1" => "http://localhost:4000/assets/app-a7b7a1bb2570182b93179aee6863d1bb.js?vsn=d"}, "vsn" => "2.0.0"}
2024-08-04 19:25:25 17:25:25.444 [info] Tzdata has updated the release from 2021e to 2024a
2024-08-04 19:25:27 17:25:27.521 [error] Unable to fetch events from AWS. Reason: {:http_error, 400, %{code: "IncompleteSignature", message: "Credential must have exactly 5 slash-delimited elements, e.g. keyid/date/region/service/term, got 'bllV7h0IQ/EEtAOZSbOOpQQOHsszvmFgpM+5IXMF/20240804/eu-central-1/sqs/aws4_request'", type: "Sender", detail: nil, request_id: "bd3de966-5c23-5c0b-813f-b64fe7e4ebac"}}
2024-08-04 19:25:32 17:25:32.589 [error] Unable to fetch events from AWS. Reason: {:http_error, 400, %{code: "IncompleteSignature", message: "Credential must have exactly 5 slash-delimited elements, e.g. keyid/date/region/service/term, got 'bllV7h0IQ/EEtAOZSbOOpQQOHsszvmFgpM+5IXMF/20240804/eu-central-1/sqs/aws4_request'", type: "Sender", detail: nil, request_id: "a036ac44-53cc-502a-86f1-f36de800e78d"}}
```

**Comment by liltechnomancer** [comment]
@tobkle the error you are getting locally is from not having the "PHX_HOST" environment variable set to localhost

You need to add all the environment variables described here. https://docs.double-zero.cloud/guides/getting-started#docker-container

As far as the fly issue, I am very confused with that as it looks like it is trying to build it without the docker container which would not be correct. You can probably get passed it by running `mix deps.get` but I would try to find out what is causing fly to not use docker there.

**Comment by liltechnomancer** [comment]
@tobkle The email confirmation issue is a known bug I am tackling. https://github.com/technomancy-dev/00/issues/12
You are fine to ignore it now as it actually has no affect on the account. 

As for the bad request, is there any difference from the request in production vs the successful local ones?

**Comment by tobkle** [comment]
It was not a test and production environment but community and pro version. Both are in local docker containers only.
Of course, I had only one of the containers running at a time because of the port 4000.

These were the calls: 
The first post below was the one to the Pro Version, and it always fails.
The second post below was the one to the Community Version, which worked the very first time. In all calls afterwards, it hasn't worked any more. Tried it with container stops and restarts. Also deleting containers and starting new containers with new api keys. No success.

Do I do anything wrong?

I can see in AWS SES, that it was not deliverable...
<img width="817" alt="2024-08-09_12-58-57" src="https://github.com/user-attachments/assets/b53773af-7ee5-4796-8ea7-687f81c8c671">

SQS Queue is empty, also SNS and CloudWatch cannot see anything.

```js
# Sending Test to Pro-Version E-Mail
POST http://localhost:4000/api/emails
Content-Type: application/json
Authorization: Bearer X57...

{
  "html": "<h1>Hello world!</h1>",
  "subject": "Can your email service track multiple recipients?",
  "from": "no-reply@email.com",
  "to": ["no-reply@email.com"]
}
```

...showing:
```js
HTTP/1.1 400 Bad Request
date: Fri, 09 Aug 2024 10:23:31 GMT
content-length: 31
vary: accept-encoding
content-encoding: gzip
content-type: text/html; charset=utf-8
cache-control: max-age=0, private, must-revalidate

Bad Request
```
In the docker container logs:
```log
2024-08-09 12:23:05 10:23:05.684 [info] Migrations already up
2024-08-09 12:23:05 10:23:05.816 [info] Running Phoenix00Web.Endpoint with Bandit 1.5.2 at :::4000 (http)
2024-08-09 12:23:05 10:23:05.819 [info] Access Phoenix00Web.Endpoint at https://localhost
2024-08-09 12:23:06 10:23:06.158 request_id=F-oIWRckJPNYmaYAAACn [info] GET /
2024-08-09 12:23:06 10:23:06.212 request_id=F-oIWRckJPNYmaYAAACn [info] Sent 302 in 54ms
2024-08-09 12:23:06 10:23:06.220 request_id=F-oIWRrqgPhYmaYAAAEH [info] GET /users/log_in
2024-08-09 12:23:06 10:23:06.224 request_id=F-oIWRrqgPhYmaYAAAEH [info] Sent 302 in 3ms
2024-08-09 12:23:06 10:23:06.229 request_id=F-oIWRtwllRYmaYAAAEn [info] GET /emails
2024-08-09 12:23:06 10:23:06.316 request_id=F-oIWRtwllRYmaYAAAEn [info] Sent 200 in 86ms
2024-08-09 12:23:06 10:23:06.408 [info] CONNECTED TO Phoenix.LiveView.Socket in 3ms
2024-08-09 12:23:06   Transport: :websocket
2024-08-09 12:23:06   Serializer: Phoenix.Socket.V2.JSONSerializer
2024-08-09 12:23:06   Parameters: %{"_csrf_token" => "IhgoCxxxxxlJBiQoRDp3JQQJIBoUmYo1FuTTdViz4ApiU", "_live_referer" => "undefined", "_mounts" => "0", "_track_static" => %{"0" => "http://localhost:4000/assets/app-845f681896bfb93bd67ca579bf058f7f.css?vsn=d", "1" => "http://localhost:4000/assets/app-8e688d63e55419c3891fb0b67722982d.js?vsn=d"}, "vsn" => "2.0.0"}
2024-08-09 03:16:16 qemu: uncaught target signal 11 (Segmentation fault) - core dumped
2024-08-09 12:23:32 10:23:32.044 request_id=F-oIXx4bgen43cUAAAXI [info] POST /api/emails
```

Here the call to the community version, which worked the first time, but not any more.

```js
# Sending Test to Community-Version E-Mail
POST http://localhost:4000/api/emails
Content-Type: application/json
Authorization: Bearer h5s....

{
  "html": "<h1>Hello world!</h1>",
  "subject": "Can your email service track multiple recipients?",
  "from": "no-reply@email.com",
  "to": ["no-reply@email.com"]
}
```

```js
HTTP/1.1 400 Bad Request
date: Fri, 09 Aug 2024 10:30:57 GMT
content-length: 31
vary: accept-encoding
content-encoding: gzip
content-type: text/html; charset=utf-8
cache-control: max-age=0, private, must-revalidate

Bad Request
```

```log
2024-08-09 12:29:44   Transport: :websocket
2024-08-09 12:29:44   Serializer: Phoenix.Socket.V2.JSONSerializer
2024-08-09 12:29:44   Parameters: %{"_csrf_token" => "IC97GDtGbxxxxwRqZSc1dgMoRHQkuJ6Jw6xxxxwY-NrCkjtBR", "_live_referer" => "undefined", "_mounts" => "0", "_track_static" => %{"0" => "http://localhost:4000/assets/app-f31482054921925fd908594c96a609e0.css?vsn=d", "1" => "http://localhost:4000/assets/app-a7b7a1bb2570182b93179aee6863d1bb.js?vsn=d"}, "vsn" => "2.0.0"}
2024-08-09 12:29:47 10:29:47.095 request_id=F-oItnD1thq57WoAAAWj [info] POST /users/log_in
2024-08-09 12:29:47 10:29:47.675 request_id=F-oItnD1thq57WoAAAWj [info] Sent 302 in 579ms
2024-08-09 12:29:47 10:29:47.681 request_id=F-oItpPbvEu57WoAAAXD [info] GET /emails
2024-08-09 12:29:47 10:29:47.709 request_id=F-oItpPbvEu57WoAAAXD [info] Sent 200 in 27ms
2024-08-09 12:29:47 10:29:47.753 [info] CONNECTED TO Phoenix.LiveView.Socket in 167µs
2024-08-09 12:29:47   Transport: :websocket
2024-08-09 12:29:47   Serializer: Phoenix.Socket.V2.JSONSerializer
2024-08-09 12:29:47   Parameters: %{"_csrf_token" => "GkV9KStWBTMQFxQZKRMVFi8HN1FkV2kkPxxxxxERWGgAGxxxF", "_live_referer" => "undefined", "_mounts" => "0", "_track_static" => %{"0" => "http://localhost:4000/assets/app-f31482054921925fd908594c96a609e0.css?vsn=d", "1" => "http://localhost:4000/assets/app-a7b7a1bb2570182b93179aee6863d1bb.js?vsn=d"}, "vsn" => "2.0.0"}
2024-08-09 12:30:58 10:30:58.406 request_id=F-oIxwtWgquLMjAAAADp [info] POST /api/emails
```

**Comment by tobkle** [comment]
Heureka, I finally got the community version working!

The problem was neither in 00 nor AWS. 
The problem was with the VSCODE REST Client, which I always use and never had an issue with it - so far.
Today, I tried the VSCODE Thunder Client like Web Dev Cody did in this youtube video and everything was working immediately.
I compared the calls of both of them. It is the same call. I cannot see any difference.
Haven't expected a problem at that point.
